{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "MODENCODE_DIR = '../../reference-data/modencode-data/dm6/'\n",
    "DAMID_DIR = 'damid_overlap'\n",
    "all_insertions = '0.5.21.filtered_calls_prosgfp_guts.gff'\n",
    "INSERTION_SET = 'ProsGFP guts'\n",
    "GENOME_FAI = '../../reference-data/dmel-all-chromosome-r6.17.fasta.fai'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pybedtools\n",
    "from collections import defaultdict\n",
    "import re\n",
    "from tempfile import TemporaryDirectory\n",
    "from adjustText import adjust_text\n",
    "%matplotlib inline\n",
    "\n",
    "\n",
    "def despine(ax):\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['bottom'].set_visible(False)\n",
    "    ax.spines['left'].set_visible(False)\n",
    "    return ax\n",
    "\n",
    "\n",
    "def set_shared_ylabel(a, ylabel, labelpad = 0.01):\n",
    "    \"\"\"Set a y label shared by multiple axes\n",
    "    Parameters\n",
    "    ----------\n",
    "    a: list of axes\n",
    "    ylabel: string\n",
    "    labelpad: float\n",
    "        Sets the padding between ticklabels and axis label\"\"\"\n",
    "\n",
    "    f = a[0].get_figure()\n",
    "    f.canvas.draw() #sets f.canvas.renderer needed below\n",
    "\n",
    "    # get the center position for all plots\n",
    "    top = a[0].get_position().y1\n",
    "    bottom = a[-1].get_position().y0\n",
    "\n",
    "    # get the coordinates of the left side of the tick labels \n",
    "    x0 = 1\n",
    "    for at in a:\n",
    "        at.set_ylabel('') # just to make sure we don't and up with multiple labels\n",
    "        bboxes, _ = at.yaxis.get_ticklabel_extents(f.canvas.renderer)\n",
    "        bboxes = bboxes.inverse_transformed(f.transFigure)\n",
    "        xt = bboxes.x0\n",
    "        if xt < x0:\n",
    "            x0 = xt\n",
    "    tick_label_left = x0\n",
    "\n",
    "    # set position of label\n",
    "    a[-1].set_ylabel(ylabel)\n",
    "    a[-1].yaxis.set_label_coords(tick_label_left - labelpad,(bottom + top)/2, transform=f.transFigure)\n",
    "\n",
    "\n",
    "def get_figsize_ncols_nrows(n, ncols=None, nrows=None, per_ax_size=5):\n",
    "    if ncols:\n",
    "        nrows = int(n / ncols)\n",
    "        if float(n) / ncols != nrows:\n",
    "            nrows += 1\n",
    "    else:\n",
    "        ncols = int(n / nrows) + 1\n",
    "        if float(n) / nrows != ncols:\n",
    "            ncols += 1\n",
    "    return (nrows * per_ax_size, ncols * per_ax_size), ncols, nrows\n",
    "\n",
    "\n",
    "# fig, ax = plt.subplots(figsize=(10, 10))\n",
    "def plot_volcano(df, insertion_label, tissue, p_value_cutoff=10, imageext='pdf'):\n",
    "    df = df[(df['log2_enrichment'] > -10) & (df['log2_enrichment'] < 10)]\n",
    "    df['Significant'] = (df['-log10 p-value'] > p_value_cutoff)\n",
    "    g = sns.lmplot(x='log2_enrichment', y='-log10 p-value', hue='Significant', data=df, fit_reg=False, size=10, aspect=1.6)\n",
    "    ax = plt.gca()\n",
    "    for index, row in df.iterrows():\n",
    "        texts = []\n",
    "        if row['-log10 p-value'] > p_value_cutoff:\n",
    "            x = row['log2_enrichment']\n",
    "            y = row['-log10 p-value']\n",
    "            text = row['Assay Factor']\n",
    "            texts.append(ax.text(x, y, text, horizontalalignment='left', size='small', color='black'))\n",
    "            adjust_text(texts, arrowprops=dict(arrowstyle=\"-\", color='k', lw=1), force_text=(0.2, 1))\n",
    "    title = \"Volcanoplot (%s) and somatic %s insertions\" % (tissue, insertion_label)\n",
    "    ax.set_title(title)\n",
    "    despine(ax)\n",
    "    ax.figure.savefig(\"%s - %s.%s\" % (INSERTION_SET, title, imageext))\n",
    "    \n",
    "\n",
    "def get_enrichments(query, target_dir, subset=None):\n",
    "    genomic_features = [os.path.join(target_dir, p) for p in os.listdir(target_dir) if p.endswith('.bed')]\n",
    "    bedtool = pybedtools.BedTool(query).sort()\n",
    "    if subset:\n",
    "        df = bedtool.to_dataframe()\n",
    "        if 'feature' in df.columns:\n",
    "            column = 'feature'\n",
    "        elif 'name' in df.columns:\n",
    "            column = 'name'\n",
    "        else:\n",
    "            raise AssertionError(\"Dataset doesn't seem to be GFF or BED.\")\n",
    "        bedtool = pybedtools.BedTool.from_dataframe(df[df[column] == subset])\n",
    "\n",
    "    result = defaultdict(dict)\n",
    "    for f in genomic_features:\n",
    "        basename = os.path.basename(f)\n",
    "        with open(f) as b:\n",
    "            r = bedtool.fisher(b, g=GENOME_FAI)\n",
    "        result[basename]['ratio'] = r.ratio\n",
    "        result[basename]['p_left'] = r.left_tail\n",
    "        result[basename]['p_right'] = r.right_tail\n",
    "        result[basename]['two_tail'] = r.two_tail\n",
    "\n",
    "    df = pd.DataFrame.from_dict(result).transpose()\n",
    "    df['index'] = [s.split('_')[0] for s in df.index]\n",
    "    df['log2_enrichment'] = np.log2(df['ratio'] + 0.00001)\n",
    "    # df = df[(df['log2_enrichment'] > -10) & (df['log2_enrichment'] < 10)]\n",
    "    df['-log10 p-value'] = -np.log10(df['two_tail'])\n",
    "    k = ['Assay Factor', 'genotype', 'Tissue', 'Type', 'replicate', 'enriched', 'project_id']\n",
    "    d = {f: dict(zip(k, f.split('.')[0].split('_'))) for f in df.index}\n",
    "    full_df = pd.DataFrame.from_dict(d).transpose().merge(df, right_index=True, left_index=True)\n",
    "    full_df['Assay Factor'] = full_df['Assay Factor'].str.replace('RNA-polymerase-II-CTD-repeat-YSPTSPS', 'PolII')\n",
    "    full_df['Assay Factor'] = full_df['Assay Factor'].str.replace('RNA-polymerase-II', 'PolII')\n",
    "    return full_df\n",
    "\n",
    "\n",
    "def mean_for_group(df):\n",
    "    df = df[df['enriched'] == 'enriched']\n",
    "    by_group = df.groupby(['Assay Factor', 'Tissue', 'genotype', 'project_id'])['Type'].count().reset_index()\n",
    "    # At least two replicates\n",
    "    by_group = by_group[by_group.Type > 1]\n",
    "    # Get the project ids with the most common project ids\n",
    "    project_ids = by_group.groupby(['Assay Factor', 'Tissue', 'genotype'])['project_id'].max().values\n",
    "    df = df[df.project_id.isin(project_ids)]\n",
    "    df = df.groupby(['Assay Factor', 'genotype', 'Tissue']).agg({'-log10 p-value': np.mean, 'log2_enrichment': np.mean}).reset_index()\n",
    "    df.index = df['Assay Factor']\n",
    "    return df\n",
    "\n",
    "\n",
    "def enriched_for_tissue(df, tissue='Adult', genotype='wt'):\n",
    "    return df[(df['genotype'] == genotype) & (df['Tissue'] == tissue)]\n",
    "\n",
    "\n",
    "def plot_heatmap_and_barchart(df, insertion_label='rover', tissue='Adult', imgext='pdf'):\n",
    "    palette = sns.cubehelix_palette(50, hue=0.05, rot=0, light=2, dark=0, as_cmap=True)\n",
    "    df = df[df.log2_enrichment != np.inf]\n",
    "    sign = (df['log2_enrichment'] > 0).astype(int)\n",
    "    sign[sign == 0] = -1\n",
    "    df['signed-log10 p-value'] = df['-log10 p-value'] * sign\n",
    "    df = df.sort_values(['signed-log10 p-value'], ascending=False)\n",
    "    to_plot = df[['-log10 p-value']]\n",
    "    height = 0.2 * len(to_plot)\n",
    "    fig = plt.figure(figsize=(5, height), constrained_layout=True)\n",
    "    widths = [1, 4]\n",
    "    spec = fig.add_gridspec(ncols=2, nrows=1, width_ratios=widths)\n",
    "    ax1 = fig.add_subplot(spec[0])\n",
    "    ax2 = fig.add_subplot(spec[1])\n",
    "\n",
    "    sns.heatmap(to_plot, center=0, annot=True, cbar=False, cmap=palette, ax=ax1)\n",
    "    to_plot = df[['log2_enrichment', '-log10 p-value']]\n",
    "    to_plot['Assay Factor'] = to_plot.index\n",
    "    sns.barplot(y='Assay Factor', x='log2_enrichment', data=to_plot, ax=ax2, color='k')\n",
    "    ax2.set_yticklabels('')\n",
    "    ax2.set_ylabel('')\n",
    "    ax2.set_yticks([])\n",
    "    despine(ax2)\n",
    "    title = \"Correlation of %s insertions with Modencode tracks for %s\" % (insertion_label, tissue)\n",
    "    fig.suptitle(title)\n",
    "    fig.savefig(\"%s - %s.%s\" % (INSERTION_SET, title, imgext), bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "copia_df = get_enrichments(query=all_insertions, target_dir=MODENCODE_DIR, subset='copia')\n",
    "rover_df = get_enrichments(query=all_insertions, target_dir=MODENCODE_DIR, subset='rover')\n",
    "i_element_df = get_enrichments(query=all_insertions, target_dir=MODENCODE_DIR, subset='I-element')\n",
    "two_nine_seven_df = get_enrichments(query=all_insertions, target_dir=MODENCODE_DIR, subset='297')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'rover_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-633ee18dc0fa>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdata_frames\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mrover_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopia_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi_element_df\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0minsertions_labels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'rover'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'copia'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'I-element'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'297'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mtissues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'Adult'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Larvae'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Embryo'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'S2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Kc'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0minsertion_label\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minsertions_labels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata_frames\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'rover_df' is not defined"
     ]
    }
   ],
   "source": [
    "data_frames = [rover_df, copia_df, i_element_df, two_nine_seven_df]\n",
    "insertions_labels = ['rover', 'copia', 'I-element', '297']\n",
    "tissues = ['Adult', 'Larvae', 'Embryo', 'S2', 'Kc']\n",
    "\n",
    "for insertion_label, df in zip(insertions_labels, data_frames):\n",
    "    for tissue in tissues:\n",
    "        enriched_df = enriched_for_tissue(mean_for_group(df), tissue=tissue)\n",
    "        if len(enriched_df) < 5:\n",
    "            continue\n",
    "        title = \"Correlation of %s insertions with Modencode tracks for %s\" % (insertion_label, tissue)\n",
    "        enriched_df.to_csv(\"%s - %s.tsv\" % (INSERTION_SET, title), sep='\\t', index=None)\n",
    "        plot_volcano(enriched_df, insertion_label, tissue, p_value_cutoff=5)\n",
    "        plot_heatmap_and_barchart(enriched_df, insertion_label=insertion_label, tissue=tissue)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
