{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "all_insertions = '0.5.21.filtered_calls_prosgfp_guts.gff'\n",
    "FEATURE_DIR = '.'\n",
    "INSERTION_SET = 'ProsGFP guts'\n",
    "GENOME_FAI = '../../../reference-data/dmel-all-chromosome-r6.17.fasta.fai'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pybedtools\n",
    "from collections import defaultdict\n",
    "import re\n",
    "from tempfile import TemporaryDirectory\n",
    "from adjustText import adjust_text\n",
    "%matplotlib inline\n",
    "\n",
    "\n",
    "def despine(ax):\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['bottom'].set_visible(False)\n",
    "    ax.spines['left'].set_visible(False)\n",
    "    return ax\n",
    "\n",
    "\n",
    "def set_shared_ylabel(a, ylabel, labelpad = 0.01):\n",
    "    \"\"\"Set a y label shared by multiple axes\n",
    "    Parameters\n",
    "    ----------\n",
    "    a: list of axes\n",
    "    ylabel: string\n",
    "    labelpad: float\n",
    "        Sets the padding between ticklabels and axis label\"\"\"\n",
    "\n",
    "    f = a[0].get_figure()\n",
    "    f.canvas.draw() #sets f.canvas.renderer needed below\n",
    "\n",
    "    # get the center position for all plots\n",
    "    top = a[0].get_position().y1\n",
    "    bottom = a[-1].get_position().y0\n",
    "\n",
    "    # get the coordinates of the left side of the tick labels \n",
    "    x0 = 1\n",
    "    for at in a:\n",
    "        at.set_ylabel('') # just to make sure we don't and up with multiple labels\n",
    "        bboxes, _ = at.yaxis.get_ticklabel_extents(f.canvas.renderer)\n",
    "        bboxes = bboxes.inverse_transformed(f.transFigure)\n",
    "        xt = bboxes.x0\n",
    "        if xt < x0:\n",
    "            x0 = xt\n",
    "    tick_label_left = x0\n",
    "\n",
    "    # set position of label\n",
    "    a[-1].set_ylabel(ylabel)\n",
    "    a[-1].yaxis.set_label_coords(tick_label_left - labelpad,(bottom + top)/2, transform=f.transFigure)\n",
    "\n",
    "\n",
    "def get_figsize_ncols_nrows(n, ncols=None, nrows=None, per_ax_size=5):\n",
    "    if ncols:\n",
    "        nrows = int(n / ncols)\n",
    "        if float(n) / ncols != nrows:\n",
    "            nrows += 1\n",
    "    else:\n",
    "        ncols = int(n / nrows) + 1\n",
    "        if float(n) / nrows != ncols:\n",
    "            ncols += 1\n",
    "    return (nrows * per_ax_size, ncols * per_ax_size), ncols, nrows\n",
    "\n",
    "\n",
    "# fig, ax = plt.subplots(figsize=(10, 10))\n",
    "def plot_volcano(df, insertion_label, tissue, p_value_cutoff=10, imageext='pdf'):\n",
    "    df = df[(df['log2_enrichment'] > -10) & (df['log2_enrichment'] < 10)]\n",
    "    df['Significant'] = (df['-log10 p-value'] > p_value_cutoff)\n",
    "    g = sns.lmplot(x='log2_enrichment', y='-log10 p-value', hue='Significant', data=df, fit_reg=False, size=10, aspect=1.6)\n",
    "    ax = plt.gca()\n",
    "    for index, row in df.iterrows():\n",
    "        texts = []\n",
    "        if row['-log10 p-value'] > p_value_cutoff:\n",
    "            x = row['log2_enrichment']\n",
    "            y = row['-log10 p-value']\n",
    "            text = row['Assay Factor']\n",
    "            texts.append(ax.text(x, y, text, horizontalalignment='left', size='small', color='black'))\n",
    "            adjust_text(texts, arrowprops=dict(arrowstyle=\"-\", color='k', lw=1), force_text=(0.2, 1))\n",
    "    title = \"Volcanoplot (%s) and somatic %s insertions\" % (tissue, insertion_label)\n",
    "    ax.set_title(title)\n",
    "    despine(ax)\n",
    "    ax.figure.savefig(\"%s - %s.%s\" % (INSERTION_SET, title, imageext))\n",
    "    \n",
    "\n",
    "def get_enrichments(query, target_dir, subset=None):\n",
    "    genomic_features = [os.path.join(target_dir, p) for p in os.listdir(target_dir) if p.endswith('.gtf')]\n",
    "    bedtool = pybedtools.BedTool(query).sort()\n",
    "    if subset:\n",
    "        df = bedtool.to_dataframe()\n",
    "        if 'feature' in df.columns:\n",
    "            column = 'feature'\n",
    "        elif 'name' in df.columns:\n",
    "            column = 'name'\n",
    "        else:\n",
    "            raise AssertionError(\"Dataset doesn't seem to be GFF or BED.\")\n",
    "        bedtool = pybedtools.BedTool.from_dataframe(df[df[column] == subset])\n",
    "\n",
    "    result = defaultdict(dict)\n",
    "    for f in genomic_features:\n",
    "        basename = os.path.basename(f)\n",
    "        with open(f) as b:\n",
    "            r = bedtool.fisher(b, g=GENOME_FAI)\n",
    "        result[basename]['ratio'] = r.ratio\n",
    "        result[basename]['p_left'] = r.left_tail\n",
    "        result[basename]['p_right'] = r.right_tail\n",
    "        result[basename]['two_tail'] = r.two_tail\n",
    "        result[basename]['overlap'] = r.table['in -a']['in -b']\n",
    "\n",
    "    df = pd.DataFrame.from_dict(result).transpose()\n",
    "    df['index'] = [s.split('_')[0] for s in df.index]\n",
    "    df['log2_enrichment'] = np.log2(df['ratio'] + 0.00001)\n",
    "    # df = df[(df['log2_enrichment'] > -10) & (df['log2_enrichment'] < 10)]\n",
    "    df['-log10 p-value'] = -np.log10(df['two_tail'])\n",
    "    k = ['Assay Factor', 'genotype', 'Tissue', 'Type', 'replicate', 'enriched', 'project_id']\n",
    "    d = {f: dict(zip(k, f.split('.')[0].split('_'))) for f in df.index}\n",
    "    full_df = pd.DataFrame.from_dict(d).transpose().merge(df, right_index=True, left_index=True)\n",
    "    full_df['Assay Factor'] = full_df['Assay Factor'].str.replace('RNA-polymerase-II-CTD-repeat-YSPTSPS', 'PolII')\n",
    "    full_df['Assay Factor'] = full_df['Assay Factor'].str.replace('RNA-polymerase-II', 'PolII')\n",
    "    return full_df\n",
    "\n",
    "\n",
    "def mean_for_group(df):\n",
    "    df = df.groupby(['Assay Factor', 'genotype', 'Tissue', 'enriched', 'overlap']).agg({'-log10 p-value': np.mean, 'log2_enrichment': np.mean}).reset_index()\n",
    "    df.index = df['Assay Factor']\n",
    "    return df\n",
    "\n",
    "\n",
    "def enriched_for_tissue(df, tissue='Adult', genotype='wt'):\n",
    "    return df[(df['genotype'] == genotype) & (df['Tissue'] == tissue) & (df['enriched'] == 'enriched')]\n",
    "\n",
    "\n",
    "def plot_heatmap_and_barchart(df, insertion_label='rover', tissue='Adult', imgext='pdf'):\n",
    "    palette = sns.cubehelix_palette(50, hue=0.05, rot=0, light=2, dark=0, as_cmap=True)\n",
    "    df = df[df.log2_enrichment != np.inf]\n",
    "    sign = (df['log2_enrichment'] > 0).astype(int)\n",
    "    sign[sign == 0] = -1\n",
    "    df['signed-log10 p-value'] = df['-log10 p-value'] * sign\n",
    "    df = df.sort_values(['signed-log10 p-value'], ascending=False)\n",
    "    to_plot = df[['-log10 p-value']]\n",
    "    height = 0.2 * len(to_plot)\n",
    "    fig = plt.figure(figsize=(5, 2), constrained_layout=True)\n",
    "    widths = [1, 4]\n",
    "    spec = fig.add_gridspec(ncols=2, nrows=1, width_ratios=widths)\n",
    "    ax1 = fig.add_subplot(spec[0])\n",
    "    ax2 = fig.add_subplot(spec[1])\n",
    "\n",
    "    sns.heatmap(to_plot, center=0, annot=True, cbar=False, cmap=palette, ax=ax1)\n",
    "    to_plot = df[['log2_enrichment', '-log10 p-value']]\n",
    "    to_plot['Assay Factor'] = to_plot.index\n",
    "    sns.barplot(y='Assay Factor', x='log2_enrichment', data=to_plot, ax=ax2, color='k')\n",
    "    ax2.set_yticklabels('')\n",
    "    ax2.set_ylabel('')\n",
    "    ax2.set_yticks([])\n",
    "    despine(ax2)\n",
    "    title = \"Correlation of %s insertions with gene tracks for %s\" % (insertion_label, tissue)\n",
    "    fig.suptitle(title)\n",
    "    fig.savefig(\"%s - %s.%s\" % (INSERTION_SET, title, imgext), bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-52-9fa6933eae3e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mrover_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_enrichments\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mall_insertions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtarget_dir\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFEATURE_DIR\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubset\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'rover'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mi_element_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_enrichments\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mall_insertions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtarget_dir\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFEATURE_DIR\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubset\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'I-element'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mtwo_nine_seven_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_enrichments\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mall_insertions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtarget_dir\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFEATURE_DIR\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubset\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'297'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-51-9f17b2e71b54>\u001b[0m in \u001b[0;36mget_enrichments\u001b[0;34m(query, target_dir, subset)\u001b[0m\n\u001b[1;32m    101\u001b[0m         \u001b[0mbasename\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbasename\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    102\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 103\u001b[0;31m             \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbedtool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfisher\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mg\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'/Users/mvandenb/src/reference/dmel-all-chromosome-r6.17.fasta.fai'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    104\u001b[0m         \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mbasename\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'ratio'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mratio\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    105\u001b[0m         \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mbasename\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'p_left'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mleft_tail\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/jupyter/lib/python3.6/site-packages/pybedtools/bedtool.py\u001b[0m in \u001b[0;36mwrapped\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    359\u001b[0m             \u001b[0;31m# At runtime, this will parse the kwargs, convert streams to\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    360\u001b[0m             \u001b[0;31m# tempfiles if needed, and return all the goodies\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 361\u001b[0;31m             \u001b[0mcmds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtmp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstdin\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandle_kwargs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprog\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mprog\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    362\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    363\u001b[0m             \u001b[0;31m# Decide whether the output is BAM format or not.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/jupyter/lib/python3.6/site-packages/pybedtools/bedtool.py\u001b[0m in \u001b[0;36mhandle_kwargs\u001b[0;34m(self, prog, **kwargs)\u001b[0m\n\u001b[1;32m   1524\u001b[0m             \u001b[0;31m# programs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1525\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1526\u001b[0;31m                 \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0minarg2\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_collapse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minstream2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1527\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1528\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/jupyter/lib/python3.6/site-packages/pybedtools/bedtool.py\u001b[0m in \u001b[0;36m_collapse\u001b[0;34m(self, iterable, fn, trackline, in_compressed, out_compressed)\u001b[0m\n\u001b[1;32m   1413\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1414\u001b[0m                         \u001b[0mi\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcreate_interval_from_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1415\u001b[0;31m                     \u001b[0mout_\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1416\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1417\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "all_insertions_df = get_enrichments(query=all_insertions, target_dir=FEATURE_DIR)\n",
    "copia_df = get_enrichments(query=all_insertions, target_dir=FEATURE_DIR, subset='copia')\n",
    "rover_df = get_enrichments(query=all_insertions, target_dir=FEATURE_DIR, subset='rover')\n",
    "i_element_df = get_enrichments(query=all_insertions, target_dir=FEATURE_DIR, subset='I-element')\n",
    "two_nine_seven_df = get_enrichments(query=all_insertions, target_dir=FEATURE_DIR, subset='297')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "data_frames = [all_insertions_df, rover_df, copia_df, i_element_df, two_nine_seven_df]\n",
    "insertions_labels = ['all transposon classes', 'rover', 'copia', 'I-element', '297']\n",
    "tissues = ['ISC']\n",
    "\n",
    "for insertion_label, df in zip(insertions_labels, data_frames):\n",
    "    for tissue in tissues:\n",
    "        df['Tissue'] = tissue\n",
    "        df['genotype'] = 'wt'\n",
    "        df['enriched'] = 'enriched'\n",
    "        enriched_df = enriched_for_tissue(mean_for_group(df), tissue=tissue)\n",
    "        if len(enriched_df) < 5:\n",
    "            continue\n",
    "        title = \"Correlation of %s insertions with gene tracks for %s\" % (insertion_label, tissue)\n",
    "        enriched_df.to_csv(\"%s - %s.tsv\" % (INSERTION_SET, title), sep='\\t', index=None)\n",
    "        plot_volcano(enriched_df, insertion_label, tissue, p_value_cutoff=5)\n",
    "        plot_heatmap_and_barchart(enriched_df, insertion_label=insertion_label, tissue=tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
